plugin/role_set_utils.go

```diff
@@ -0,0 +1,161 @@
+// accounts removed from a RoleSet), and because changing RoleSet requires migration of stored RoleSets.
+// A utility for creating roleSetResources from a RoleSet is provided below.
+type roleSetResources struct {
+	rolesetName string
+
+	accountId *gcputil.ServiceAccountId
+	bindings  ResourceBindings
+	tokenGen  *TokenGenerator
+}
+
+func (rs *RoleSet) boundResources() *roleSetResources {
+	if rs.AccountId == nil {
+		return nil
+	}
+	return &roleSetResources{
+		accountId:   rs.AccountId,
+		rolesetName: rs.Name,
+		bindings:    rs.Bindings,
+		tokenGen:    rs.TokenGen,
+	}
+}
+
+// addWalsForRoleSetResources creates WALs to clean up a roleset's service account, bindings, and a key if needed.
+func (b *backend) addWalsForRoleSetResources(ctx context.Context, req *logical.Request, boundResources *roleSetResources) error {
+	if boundResources == nil {
+		b.Logger().Debug("skip WALs for nil roleset resources")
+		return nil
+	}
+
+	_, err := framework.PutWAL(ctx, req.Storage, walTypeAccount, &walAccount{
+		RoleSet: boundResources.rolesetName,
+		Id:      *boundResources.accountId,
+	})
+	if err != nil {
+		return errwrap.Wrapf("unable to create WAL entry to clean up service account: {{err}}", err)
+	}
+
+	for resName, roleSet := range boundResources.bindings {
+		_, err := framework.PutWAL(ctx, req.Storage, walTypeIamPolicy, &walIamPolicy{
+			RoleSet:   boundResources.rolesetName,
+			AccountId: *boundResources.accountId,
+			Resource:  resName,
+			Roles:     roleSet.ToSlice(),
+		})
+		if err != nil {
+			return errwrap.Wrapf("unable to create WAL entry to clean up service account bindings: {{err}}", err)
+		}
+	}
+
+	if boundResources.tokenGen != nil {
+		return b.addServiceAccountKeyWal(ctx, req, boundResources.rolesetName, boundResources.accountId, boundResources.tokenGen.KeyName)
+	}
+	return nil
+}
+
+// addServiceAccountKeyWal creates WAL to clean up a roleset service account key (for access tokens) if needed.
+func (b *backend) addServiceAccountKeyWal(ctx context.Context, req *logical.Request, rolesetName string, accountId *gcputil.ServiceAccountId, keyName string) error {
+	if accountId == nil {
+		b.Logger().Debug("skip adding service account key WAL for nil account")
+		return nil
+	}
+
+	_, err := framework.PutWAL(ctx, req.Storage, walTypeAccount, &walAccountKey{
+		RoleSet:            rolesetName,
+		ServiceAccountName: accountId.ResourceName(),
+		KeyName:            keyName,
+	})
+	if err != nil {
+		return errwrap.Wrapf("unable to create WAL entry to clean up service account key: {{err}}", err)
+	}
+	return nil
+}
+
+// tryDeleteRoleSetResources tries to delete GCP resources previously managed by a roleset.
+// This assumes that deletion of these resources will already be guaranteed by WALs and will return errors
+// as a list of warnings instead.
+func (b *backend) tryDeleteRoleSetResources(ctx context.Context, req *logical.Request, boundResources *roleSetResources) []string {
```

Comment:
I think we should return `(err error, warnings []string)` here. It feels weird to swallow up "failed to create HTTP client" as a warning.

Reply:
I do see why we might want to elevate the warning to an error, but I'm not sure I agree with the mixed approach.

Feedback is welcome. If we're trying to say "We succeeded in deleting/updating the roleset (intended operation) but we encountered an error while trying to clean up old resources - this will be covered by WAL", would it be better to 

- warn on any error - the actual operation succeeded and the other resources will get cleaned up eventually
- return the error - the actual roleset was created but we got an error, so if you're worried about any time period where there is unmanaged resources, you should go delete them
- err on the big things (can't create clients) but warn if we can't clean up the other errors.  

I do see why we might want to elevate the warning to an error, but I'm not sure I agree with the mixed approach, especially since the actual operation succeeded, which means recently the HTTP client was creatable

Reply:
No strong opinions from me on this one, but I'd be fine if these were just treated as errors. Granted, the user impact is low, but we treat the inability to create a roleset as an error, so problems deleted them (even if for clean-up) seem error-worthy. If desired you could also work in go-multierror, which already used elsewhere.